### PR TITLE
GH-233: Wake the consumer when the back pressure is relieved

### LINKED ISSUE #233: When fetching from partitions is paused, consumer.poll(pollTimeout) method should not be called
In [ConsumerEventLoop#PollEvent](https://github.com/reactor/reactor-kafka/blob/553aa5c92a25431a8fecce647d78b8e94b32cca5/src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java#L254) there are two options when fetching from the requesting partitions can be paused:

- [awaiting for transaction](https://github.com/reactor/reactor-kafka/blob/553aa5c92a25431a8fecce647d78b8e94b32cca5/src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java#L274)
- [back pressure](https://github.com/reactor/reactor-kafka/blob/553aa5c92a25431a8fecce647d78b8e94b32cca5/src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java#L280)

In both cases after `consumer.pause(consumer.assignment())`, `consumer.poll(pollTimeout)`method is called which will run until timeout is reached and return no records because fetching from the partitions is paused. Such behavior slows down receiver performance.

Example:
When poll timeout is 10 seconds and record request came immediately after back pressure pause, we had to wait 10 seconds before run method started again and fetching from partitions was resumed.
```
2021-07-28 17:26:15.011  DEBUG 3484 --- [-19f4ff15730e-1] r.k.r.internals.ConsumerEventLoop        : Started run method
2021-07-28 17:26:15.011  DEBUG 3484 --- [-19f4ff15730e-1] r.k.r.internals.ConsumerEventLoop        : Requested: 0
2021-07-28 17:26:15.011 DEBUG 3484 --- [-19f4ff15730e-1] r.k.r.internals.ConsumerEventLoop        : Paused - back pressure
2021-07-28 17:26:15.046 DEBUG 3484 --- [amController-10] r.k.r.internals.ConsumerEventLoop        : onRequest.toAdd 1
2021-07-28 17:26:25.011  DEBUG 3484 --- [-19f4ff15730e-1] r.k.r.internals.ConsumerEventLoop        : Finished run method
2021-07-28 17:26:25.012  DEBUG 3484 --- [-19f4ff15730e-1] r.k.r.internals.ConsumerEventLoop        : Started run method
2021-07-28 17:26:25.012  DEBUG 3484 --- [-19f4ff15730e-1] r.k.r.internals.ConsumerEventLoop        : Requested: 1
2021-07-28 17:26:25.012 DEBUG 3484 --- [-19f4ff15730e-1] r.k.r.internals.ConsumerEventLoop        : Resumed
....
```

## Expected Behavior
When fetching from partitions is paused, `consumer.poll(pollTimeout)` method should not be called.

* Reactor version(s) used: 1.3.4

### PR DESCRIPTION
**Pull request by arina-ielchiieva**
Resolves #233

### CODE REVIEW

File: src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java [hunk 1]
```diff
@@ -266,22 +270,31 @@ public void run() {
                                 toResume.removeAll(this.pausedByUser);
                                 this.pausedByUser.clear();
                                 consumer.resume(toResume);
+                                isPaused.set(false);
```
Comment (simonbasle) [inline src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java:273]: at each point `isPaused` is set to a boolean, `pausedByUs` has already been set to that same boolean. I'm not saying I'm sure this is a suitable replacement, but wondering if there's a good reason not to use `pausedByUs`? The only thing I can think of is that `pausedByUs` is set to `true` in more cases than the ones for which we want to `consumer.wakeup()` above at line 142.

**Review by simonbasle — CHANGES_REQUESTED** [review]
@arina-ielchiieva would you be able to demonstrate the issue (and thus the fix) in a unit test?

File: src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java (hunk 1 above)
Comment (garyrussell) [reply src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java:273]: Yes; we can use `pausedByUs` by pushing it up to the enclosing class.

However, I don't believe we can consider this PR until the CLA is "signed".

Comment (arina-ielchiieva) [reply src/main/java/reactor/kafka/receiver/internals/ConsumerEventLoop.java:273]: Good point, made changes to use `pausedByUs` as indicator to wake up the consumer.
Signed CLA as well.

**Comment by arina-ielchiieva** [comment]
@simonbasle  / @garyrussell thanks for the review. Applied required changes and added unit test.

**Review by simonbasle — APPROVED** [review]
LGTM. what do you think @garyrussell?

**Review by garyrussell — APPROVED** [review]

**Pull request merged**